timeboost-sequencer/src/decrypt.rs

```diff
@@ -1,37 +1,453 @@
+    if modified_ptxs.len() + modified_txs.len() != dec.len() {
+        return Err(DecryptError::State);
+    }
+    modified_ptxs
+        .into_iter()
+        .enumerate()
+        .map(|(i, m)| {
+            if let Some(ptx) = ptxs.get_mut(m) {
+                *ptx = PriorityBundle::new_with_hash(
+                    ptx.epoch(),
+                    ptx.seqno(),
+                    dec.get(i).ok_or(DecryptError::State)?.0.clone(),
+                    ptx.kid(),
+                );
+                Ok::<(), DecryptError>(())
+            } else {
+                Err(DecryptError::State)
+            }
+        })
+        .collect::<Result<Vec<_>, DecryptError>>()?;
+
+    modified_txs
+        .into_iter()
+        .enumerate()
+        .map(|(i, tx)| {
+            if let Some(tx) = txs.take(&tx) {
+                let data = dec
+                    .get(i + ptxs.len())
+                    .ok_or(DecryptError::InvalidMessage)?
+                    .0
+                    .clone();
+                let dec_tx = Transaction::new(tx.nonce(), tx.to(), data, tx.kid());
+                txs.insert(dec_tx);
+                Ok(())
+            } else {
+                Err(DecryptError::State)
+            }
+        })
+        .collect::<Result<Vec<_>, DecryptError>>()?;
+
+    Ok(())
+}
+
+impl Drop for Decrypter {
+    fn drop(&mut self) {
+        self.jh.abort()
+    }
+}
+
+type Incubator = Arc<Mutex<BTreeMap<DecShareKey, Vec<DecShare>>>>;
+
+/// Worker is responsible for "hatching" ciphertexts.
+///
+/// When ciphertexts in a round have received t+1 decryption shares
+/// the shares can be combined to decrypt the ciphertext (hatching).
+struct Worker {
+    net: Network,
+    committee: Keyset,
+    dec_sk: DecryptionKey,
+    idx2cid: BiMap<usize, Nonce>,
+    cid2ct: BiMap<Nonce, Ciphertext>,
+    shares: Incubator,
+}
+
+impl Worker {
+    pub fn new(net: Network, committee: Keyset, dec_sk: DecryptionKey) -> Self {
+        Self {
+            net,
+            committee,
+            dec_sk,
+            idx2cid: BiMap::new(),
+            cid2ct: BiMap::new(),
+            shares: Incubator::default(),
+        }
+    }
+
+    pub async fn go(
+        mut self,
+        mut enc_rx: Receiver<(RoundNumber, Vec<EncryptedItem>)>,
+        dec_tx: Sender<(RoundNumber, Vec<DecryptedItem>)>,
+    ) {
+        let mut r = RoundNumber::new(0);
+        loop {
+            tokio::select! {
+                // received batch of decryption shares from remote node.
+                Ok((pubkey, bytes)) = self.net.receive() => {
+                    if let Ok(s) = bincode::deserialize::<ShareInfo>(&bytes) {
+                        r = s.round();
+                        if let Err(e) = self.insert_shares(s) {
+                            warn!("failed to insert shares from remote: {:?}", e);
+                        }
+
+                    } else {
+                        warn!("failed to deserialize share from: {}", pubkey);
```

Comment:
```suggestion
                        warn!("failed to deserialize share from: {}", pubkey);
                        continue;
```

Does it make sense to enter `hatch` after this error?

Reply:
Good catch. Indeed, no need to hatch if no shares have been added to the incubator.